Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp

```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ *
+ * Atomically increments the reference count of @user_script_message by one.
+ *
+ * Returns: the @user_script_message passed in.
+ *
+ * Since: 2.38
+ */
+WebKitUserScriptMessage* webkit_user_script_message_ref(WebKitUserScriptMessage* userScriptMessage)
+{
+    g_return_val_if_fail(userScriptMessage, nullptr);
+    g_atomic_int_inc(&userScriptMessage->referenceCount);
+    return userScriptMessage;
+}
+
+/**
+ * webkit_user_script_message_unref:
+ * @user_script_message*: A #WebKitUserScriptMessage
+ *
+ * Atomically decrements the reference count of @user_script_message by one.
+ *
+ * If the reference count drops to 0, all the memory allocated by the
+ * #WebKitUserScriptMessage is released. This function is MT-safe and may
+ * be called from any thread.
+ *
+ * Since: 2.38
+ */
+void webkit_user_script_message_unref(WebKitUserScriptMessage* userScriptMessage)
+{
+    g_return_if_fail(userScriptMessage);
+    if (g_atomic_int_dec_and_test(&userScriptMessage->referenceCount)) {
+        userScriptMessage->~WebKitUserScriptMessage();
+        fastFree(userScriptMessage);
+    }
+}
+
+WebKitUserScriptMessage* webKitUserScriptMessageCreate(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
+{
+    WebKitUserScriptMessage* userScriptMessage = static_cast<WebKitUserScriptMessage*>(fastMalloc(sizeof(WebKitUserScriptMessage)));
+    new (userScriptMessage) WebKitUserScriptMessage(serializedScriptValue, WTFMove(completionHandler));
+    return userScriptMessage;
+}
+
+/**
+ * webkit_user_script_message_get_value:
+ * @user_script_message: A #WebKitUserScriptMessage
+ *
+ * Return the message sent in by user.
+ *
+ * Returns: (transfer none): The #JSCValue sent by JavaScript code to the
+ * script message handler.
+ *
+ * Since: 2.38
+ */
+JSCValue* webkit_user_script_message_get_value(WebKitUserScriptMessage* userScriptMessage)
+{
+    return userScriptMessage->message.get();
+}
+
+/**
+ * webkit_user_script_message_reply:
+ * @user_script_message: A #WebKitUserScriptMessage
+ * @reply_value: Reply value of the provided script
+ *
+ * Provide the reply to the user provided script.
+ *
+ * Since: 2.38
+ */
+void webkit_user_script_message_reply(WebKitUserScriptMessage* message, JSCValue* replyValue)
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(JSC_IS_VALUE(replyValue));
+    message->replyValue = replyValue;
+}
+
+/**
+ * webkit_user_script_message_error:
+ * @user_script_message: A #WebKitUserScriptMessage
+ * @error: (transfer full): An error to return as the result of running the user script.
+ *
+ * Provide the error thrown when processing user provided script.
+ *
+ * Since: 2.38
+ */
+void webkit_user_script_message_error(WebKitUserScriptMessage* message, GError* error)
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(error != nullptr);
+    message->replyError.reset(error);
```

Comment:
Same here: this needs to call the `completionHandler` with the error.

One possible way of converting the `GError` to a string can be:

```cpp
GUniquePtr<char> formattedError(g_strdup_printf("%s (%i): %s", g_quark_to_string(error->domain), error->code, error->message));